driver-core/src/main/com/mongodb/internal/connection/ClientMetadataHelper.java

```diff
@@ -102,69 +80,95 @@ private static boolean nameMatches(final String name, final String... prefixes)
+        MongoDriverInformation baseDriverInfor = getDriverInformation(null);
+        // required fields:
+        tryWithLimit(client, d -> {
+            putAtPath(d, "driver.name", listToString(baseDriverInfor.getDriverNames()));
+            putAtPath(d, "driver.version", listToString(baseDriverInfor.getDriverVersions()));
+        });
+        tryWithLimit(client, d -> putAtPath(d, "os.type", getOperatingSystemType(getOperatingSystemName())));
+        // full driver information:
+        MongoDriverInformation fullDriverInfo = getDriverInformation(mongoDriverInformation);
+        tryWithLimit(client, d -> {
+            putAtPath(d, "driver.name", listToString(fullDriverInfo.getDriverNames()));
+            putAtPath(d, "driver.version", listToString(fullDriverInfo.getDriverVersions()));
+        });
+        // optional fields:
+        Environment environment = getEnvironment();
+        tryWithLimit(client, d -> putAtPath(d, "env.name", environment.getName()));
+        tryWithLimit(client, d -> putAtPath(d, "os.name", getOperatingSystemName()));
+        tryWithLimit(client, d -> putAtPath(d, "os.architecture", getProperty("os.arch", "unknown")));
+        tryWithLimit(client, d -> putAtPath(d, "os.version", getProperty("os.version", "unknown")));
+        tryWithLimit(client, d -> putAtPath(d, "env.timeout_sec", environment.getTimeoutSec()));
+        tryWithLimit(client, d -> putAtPath(d, "env.memory_mb", environment.getMemoryMb()));
+        tryWithLimit(client, d -> putAtPath(d, "env.region", environment.getRegion()));
+        tryWithLimit(client, d -> putAtPath(d, "env.url", environment.getUrl()));
+        tryWithLimit(client, d -> putAtPath(d, "platform", listToString(baseDriverInfor.getDriverPlatforms())));
+        tryWithLimit(client, d -> putAtPath(d, "platform", listToString(fullDriverInfo.getDriverPlatforms())));
 
-        if (mongoDriverInformation != null) {
-            addDriverInformation(mongoDriverInformation, document);
-        }
+        cachedClientMetadataDocument = client;
+        return client;
+    }
 
-        if (clientMetadataDocumentTooLarge(document)) {
-            // first try: remove the three optional fields in the 'os' document, if it exists (may not if the security manager is configured
-            // to disallow access to system properties)
-            BsonDocument operatingSystemDocument = document.getDocument(OS_FIELD, null);
-            if (operatingSystemDocument != null) {
-                operatingSystemDocument.remove(OS_VERSION_FIELD);
-                operatingSystemDocument.remove(OS_ARCHITECTURE_FIELD);
-                operatingSystemDocument.remove(OS_NAME_FIELD);
-            }
-            if (operatingSystemDocument == null || clientMetadataDocumentTooLarge(document)) {
-                // second try: remove the optional 'platform' field
-                document.remove(PLATFORM_FIELD);
-                if (clientMetadataDocumentTooLarge(document)) {
```

Comment:
The original logic would remove more fields, if the document was too large. The new logic will not add items to the result if over the limit, which is easier to follow (the additions and order-of-priority are in the same code location).